magenta-lib/src/main/scala/magenta/graph/Graph.scala

```diff
@@ -66,6 +96,50 @@ case class Graph[T](edges: Set[Edge[T]]) {
     Graph(newEdges)
   }
 
+  /**
+    * Flat map keeps the shape of this graph whilst mapping each node to a graph. This means one can replace the Start,
+    * End or Mid nodes with multiple nodes. The incoming and outgoing edges of each node of this graph will be replaced
+    * by edges that map to the start and end nodes of the graph returned from f.
+    *
+    * An identity mapping for flatMap would be start and end nodes to empty map and mid nodes to a single value graph
+    * containing the value of the node.
+    *
+    * It is not possible to map a mid node to an empty node.
+    * @param f the function to apply to each element
+    * @return a new graph with the graphs substituted
+    */
+  def flatMap[R](f: Node[T] => Graph[R]): Graph[R] = {
+    val nodeToGraphMap: Map[Node[T], Graph[R]] = nodes.map(n => n -> f(n)).toMap
+    // find all of the edges that we won't adjoin to another graph
+    val allInternalEdges = nodeToGraphMap.flatMap{
+      case (StartNode, graph) => graph.edges -- graph.incoming(EndNode)
+      case (MidNode(_), graph) =>
+        assert(!graph.isEmpty, "It is not possible to flatMap a MidNode to an empty graph")
```

Comment:
Yeah - try to actually do that though. Maybe it's easier than I think but I chose to do this because it's significantly simpler not to be able to remove nodes.